rfcs/000-consolidate-amundsen-microservices.md

```diff
@@ -0,0 +1,154 @@
+
+*This change results in faster local development, one PR to fix a bug or a feature, no dependency hell as we are facing today, hence attract more contributors.*
+
+#### 3: Customization & Deployments
+One of the frequently mentioned topics re Amundsen is the complexity of the architecture, and the efforts required to customize Amundsen. Having multiple repositories results in multiple components to install, customize, manage and fix.
+
+*This change will make the customization much easier and the deployment will be pretty straightforward, with only 2 components. No multilevel docker files, easy maintenance of the infrastructure.* 
+
+Custom endpoints will still be injectable the same way we are doing today. We will write a blog/tutorial around each customization.
+
+#### 4: Code Duplication
+From requirements to config files, models, helper functions, exceptions handling, CI/CD pipelines, Docker files, and other repository management like licenses, PR/Issues templates, etc., is pretty much the same in all these microservices. Change in one repository deviates that repository from others at the moment. 
+
+*This change will move all the above into a single repository making it easy to change anything without keeping track of multiple repositories, and redundant code.* 
+
+Having one config file does NOT mean we can not have the incubator integration. It will be the same process as of today i.e., a separate incubation repo, which implements the required methods, and then install that as a requirement and set that as a proxy within the config file.
+
+
+## Transition Path
+Most of the code is already there and is spread across multiple repositories. I can see this happen in multiple phases.
+
+#### Phase I:
+Merge `amundsenmetadatalibrary` and `amundsensearchlibrary`. This should not take too long and will be easy to maintain the same URL patterns for each under separate directories within one repository. 
+The new repository will be named `amundsen`. During this period the `amundsenfrontend` repository will be unchanged and will work as it is. 
+
+Metadata and Search will have their own Flask blueprints. 
+- /api/metadata/
+- /api/search/
+
+During this phase, we will freeze the codebase for `amundsenmetadatalibrary` and `amundsensearchlibrary` repositories for any new features. Security patches will still be approved and merged.
+
+A very high level of how the directories will look like after this phase is below:
+
+<img src="../assets/30/phase1.png" alt="drawing" width="300"/>
+
+#### Phase II:
+In this phase, we’ll deprecate the `amundsenfrontendlibrary` and `amundsencommon` repositories/packages. 
+We’ll move the react application as-is in the `amundsen` repository. 
+Next will be to deprecate the frontend codebase completely and call the metadata and search endpoints directly from the frontend react application. This will completely remove the frontend Flask codebase. 
```

Comment:
This will change the security model, in that the metadata and search services have mutation endpoints that aren't OK to be exposed unprotected to web. I think people typically secure it by firewalling it off except for approved services, so by virtue of limited functionality from the front-end, it's not a problem.

Should this be figured out now, or in a future RFC?

Reply:
It is also sometimes nice to be able to define "backend-for-frontend style" APIs; that is, API endpoints that are somewhat specific to the UI in question, rather than being generally-useful unopinionated UIs. I suppose that could be supported with a separate blueprint?